c-layer/include/lib.h

```diff
@@ -51,24 +51,69 @@ enum RESOLVABLE_KIND {
   RESOLVABLE_PATTERN,
 };
 
+
+enum RESOLVABLE_SELECTED {
+  /// resolvable won't be installed
+  NOT_SELECTED,
+  /// dependency solver select resolvable
+  /// match TransactByValue::SOLVER
+  SOLVER_SELECTED,
+  /// installation proposal selects resolvable
+  /// match TransactByValue::APPL_{LOW,HIGH} we do not need both, so we use just one value
+  APPLICATION_SELECTED,
+  /// user select resolvable for installation
+  /// match TransactByValue::USER
+  USER_SELECTED,
+};
+
 /// Marks resolvable for installation
 /// @param name resolvable name
 /// @param kind kind of resolvable
+/// @param who who do selection. If NOT_SELECTED is used, it will be empty operation.
 /// @param[out] status (will overwrite existing contents)
-void resolvable_select(const char* name, enum RESOLVABLE_KIND kind, struct Status* status) noexcept;
+void resolvable_select(const char *name, enum RESOLVABLE_KIND kind, enum RESOLVABLE_SELECTED who, struct Status *status) noexcept;
 
 /// Unselect resolvable for installation. It can still be installed as dependency.
 /// @param name resolvable name
 /// @param kind kind of resolvable
+/// @param who who do unselection. Only unselect if it is higher or equal level then who do the selection.
 /// @param[out] status (will overwrite existing contents)
-void resolvable_unselect(const char* name, enum RESOLVABLE_KIND kind, struct Status* status) noexcept;
+void resolvable_unselect(const char *name, enum RESOLVABLE_KIND kind, enum RESOLVABLE_SELECTED who, struct Status *status) noexcept;
+
+
+struct PatternNames {
+  /// names of patterns
+  const char *const *const names;
+  /// size of names array
+  unsigned size;
+};
+
+struct PatternInfo {
```

Comment:
```suggestion
/// Info from zypp::Pattern.
/// https://doc.opensuse.org/projects/libzypp/HEAD/classzypp_1_1Pattern.html
struct PatternInfo {
```

Sadly even that link does not contain better info about the contained data